### ISSUE TITLE
data resurrection in simple_decommission_node_2_test

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylla#5838]
Regression first seen in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/393/testReport/update_cluster_layout_tests/TestUpdateClusterLayout/simple_decommission_node_2_test/
Scylla version b01b11c1f3fb245757402e157fca42c51c47362b
(last passed at e93c54e83798190a0d0bbc80c11fb37dcb604132)
```
Error Message
243 != 0 : expected 0 lines got 243
Stacktrace
Traceback (most recent call last):
  File "/usr/lib64/python3.7/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib64/python3.7/unittest/case.py", line 645, in run
    testMethod()
  File "/jenkins/workspace/scylla-master/dtest-release/scylla-dtest/update_cluster_layout_tests.py", line 764, in simple_decommission_node_2_test
    self.assertEqual(len(result), 0, "expected 0 lines got %d" % len(result))
  File "/usr/lib64/python3.7/unittest/case.py", line 869, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib64/python3.7/unittest/case.py", line 862, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 243 != 0 : expected 0 lines got 243
```

Also seen in https://jenkins.scylladb.com/view/scylla-3.3/job/scylla-3.3/job/dtest-release/17/testReport/update_cluster_layout_tests/TestUpdateClusterLayout/simple_decommission_node_2_test/
Scylla version 756574d09464b54421f539fac6640313ccbe8189
Passed in f842154453b89db1d71cc8988cc44c57dcc728f3

**Comment by raphaelsc** [comment]
On Tue, Feb 18, 2020 at 10:30 AM Benny Halevy <notifications@github.com>
wrote:

> Regression first seen in
> https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/393/testReport/update_cluster_layout_tests/TestUpdateClusterLayout/simple_decommission_node_2_test/
> Scylla version b01b11c
> <https://github.com/scylladb/scylla/commit/b01b11c1f3fb245757402e157fca42c51c47362b>
> (last passed at e93c54e
> <https://github.com/scylladb/scylla/commit/e93c54e83798190a0d0bbc80c11fb37dcb604132>
> )
>
> Error Message
> 243 != 0 : expected 0 lines got 243
> Stacktrace
> Traceback (most recent call last):
>   File "/usr/lib64/python3.7/unittest/case.py", line 60, in testPartExecutor
>     yield
>   File "/usr/lib64/python3.7/unittest/case.py", line 645, in run
>     testMethod()
>   File "/jenkins/workspace/scylla-master/dtest-release/scylla-dtest/update_cluster_layout_tests.py", line 764, in simple_decommission_node_2_test
>     self.assertEqual(len(result), 0, "expected 0 lines got %d" % len(result))
>   File "/usr/lib64/python3.7/unittest/case.py", line 869, in assertEqual
>     assertion_func(first, second, msg=msg)
>   File "/usr/lib64/python3.7/unittest/case.py", line 862, in _baseAssertEqual
>     raise self.failureException(msg)
> AssertionError: 243 != 0 : expected 0 lines got 243
>
> Also seen in
> https://jenkins.scylladb.com/view/scylla-3.3/job/scylla-3.3/job/dtest-release/17/testReport/update_cluster_layout_tests/TestUpdateClusterLayout/simple_decommission_node_2_test/
> Scylla version 756574d
> <https://github.com/scylladb/scylla/commit/756574d09464b54421f539fac6640313ccbe8189>
> Passed in f842154
> <https://github.com/scylladb/scylla/commit/f842154453b89db1d71cc8988cc44c57dcc728f3>
>

That's #4446. Fix sent to mailing list as '[PATCH] sstables/cleanup: Fix
stale data being returned due to lack of cache invalidation'

**Comment by bhalevy** [comment]
But this doesn't explain the regression, does it?

As I suspected, test passes with 5e9925b9f01961e010cb5e72fba199ad6dbd9420 reverted.
It was also backported to:
- 3.3 as bdc542143e374af64b08450c79f602d0a4949b86
- 3.2 as dbf72c72b3dcb811deb388848b4c6b4d31b3d1b8
- 3.1 as 4791b726a0fb13bd82763d9b9f3264ae230cc1f8

Ref #5769

**Comment by bhalevy** [comment]
Cc @avikivity @slivne @asias 

We may want to revert  5e9925b and backports of it mentioned above and reopen #5769

**Comment by raphaelsc** [comment]
On Tue, Feb 18, 2020 at 10:40 AM Benny Halevy <notifications@github.com>
wrote:

> But this doesn't explain the regression, does it?
>
not the regression, which seems to be caused by the fix for #5769. so we
need to backport that cleanup invalidation fix of mine to every branch
#5759 fix was backported to.

**Comment by asias** [comment]
Just to be clear.

The patch "Do not invalidate cache if no sstable is added in flush_streaming_mutations #5769" itself is correct. It makes the bug "https://github.com/scylladb/scylla/issues/4446" visible. 

So we need to backport  the patch for #5769 and #4446 together.

**Comment by raphaelsc** [comment]
On Wed, Feb 19, 2020 at 12:32 AM Asias He <notifications@github.com> wrote:

> Just to be clear.
>
> The patch "Do not invalidate cache if no sstable is added in
> flush_streaming_mutations #5769
> <https://github.com/scylladb/scylla/issues/5769>" itself is correct. It
> makes the bug "#4446 <https://github.com/scylladb/scylla/issues/4446>"
> visible.
>
> So we need to backport the patch for #5769
> <https://github.com/scylladb/scylla/issues/5769> and #4446
> <https://github.com/scylladb/scylla/issues/4446> together.
>
FWIW, patch to fix #4446 was sent to mailing list and i am currently
working on its v2.

**Comment by bhalevy** [comment]
@avikivity please backport the reverting patch (https://github.com/scylladb/scylla/issues/5844)
to 3.{1,2,3} branches since @raphaelsc fix for #4446 is still in the works.
Cc @slivne

**Comment by avikivity** [comment]
Already backported everywhere.
